src/pmdas/amdgpu/localdrm.c

```diff
@@ -0,0 +1,323 @@
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#if defined(HAVE_DLFCN_H)
+#include <dlfcn.h>
+#endif
+#include "localdrm.h"
+
+#include <libdrm/amdgpu.h>
+#include <libdrm/amdgpu_drm.h>
+#include <xf86drm.h>
+
+#ifndef DSOSUFFIX
+#define DSOSUFFIX "so"
+#endif
+
+/* Looks like AMD kept PCI_VENDOR_ID_ATI for its GPU IDs */
+#define PCI_VENDOR_ID_ATI 0x1002
+
+int localDRMShutdown(drmDevicePtr devs[], uint32_t count) {
+    drmFreeDevices(devs, count);
+
+    free(devs);
+
+    return DRM_SUCCESS;
+}
+
+int localDRMDeviceGetDevices(drmDevicePtr *devs[],
+			     uint32_t *max,
+			     uint32_t *count) {
+  uint32_t amdgpu_count = 0;
+  drmDevicePtr *temp = NULL; /* Will contain all devices, including non-amd ones*/
+  drmDevicePtr *p = NULL; /* Helper to copy AMD device data */
+
+  /* First get the total count of devices */
+  int dev_count = drmGetDevices(NULL, 0);
+
+  if (dev_count <= 0) {
+      printf("No devices\n");
+      return DRM_ERROR_NOT_FOUND;
+  }
+
+  /* Allocate space to store device data */
+  temp = (drmDevicePtr *)calloc(dev_count, sizeof(drmDevicePtr));
+  if (!temp) {
+      printf("No memory\n");
+      return DRM_ERROR_MEMORY;
+  }
+
+  /* Allocate space for the devices given back to the user */
+  p = *devs = (drmDevicePtr *)calloc(dev_count, sizeof(drmDevicePtr));
+  if (!*devs) {
+      printf("No memory\n");
+      free(temp);
+      return DRM_ERROR_MEMORY;
+  }
+
+  dev_count = drmGetDevices(temp, dev_count);
+
+  if (dev_count <= 0) {
+      printf("Failed to retrieve devices\n");
+      free(temp);
+      free(*devs);
+      *devs = NULL;
+      return DRM_ERROR_NOT_FOUND;
+  }
+
+  /* Walk through the devices, and keep the AMD GPU ones */
+  for (uint32_t i = 0; i < dev_count; i++) {
+      if (temp[i]->bustype != DRM_BUS_PCI ||
+          temp[i]->deviceinfo.pci->vendor_id != PCI_VENDOR_ID_ATI)
+        continue;
+
+      int fd = -1;
+
+      // Try render node first
+      if (1 << DRM_NODE_RENDER & temp[i]->available_nodes) {
+          fd = open(temp[i]->nodes[DRM_NODE_RENDER], O_RDWR);
+      }
+
+      if (fd < 0) {
+          // Fallback to primary node
+          if (1 << DRM_NODE_PRIMARY & temp[i]->available_nodes) {
+              fd = open(temp[i]->nodes[DRM_NODE_PRIMARY], O_RDWR);
+          }
+      }
+
+      if (fd < 0)
+        continue;
+
+      /* Check the version, as it contains the driver name */
+      drmVersionPtr ver = drmGetVersion(fd);
+
+      if (!ver) {
+          close(fd);
+          continue;
+      }
+
+      if (strcmp(ver->name, "amdgpu")) {
+          drmFreeVersion(ver);
+          continue;
+      }
+
+      /* Copy the AMD GPU data */
+      memcpy(&p[amdgpu_count++], &temp[i], sizeof(drmDevicePtr));
+
+      /* Done with version */
+      drmFreeVersion(ver);
```

Comment:
Do we need to close the fd (local variable on-stack)?  Looks like it leaks an fd for each device here otherwise.